285 create initfun

GidonFrischkorn opened the pull request:
#### Summary
- add method `create_initfun` to take care of initial values for `bmmodels` that benefit from or require  initial values for sampling
- as an example I have set initial values for the `sdm`
- currently initial values are only set for intercept parameters; for effects they are sampled from a uniform(-0.1,0.1), for random effects (that is `Z` parameters) they are sampled from a uniform(-.5,.5), and covariance matrices / cholesky factors are initialized as diagonal matrices
- for now, I have implemented only minimal tests, as the setup to access the `create_initfun` is somewhat cumbersome

#### Tests

[x] Confirm that all tests passed
[x] Confirm that devtools::check() produces no errors

#### Release notes
- new method `create_initfun` for `bmmodels` that require initial values for optimal sampling

GidonFrischkorn committed:
Update helpers-model.R

GidonFrischkorn committed:
Add tests and resolve bug

GidonFrischkorn committed:
Add tests & minor fixes

GidonFrischkorn committed:
Add `create_initfun`

GidonFrischkorn committed:
Add `link_transform` function + tests

GidonFrischkorn committed:
Merge branch '288-utility-function-link-transform' into 285-create_initfun

GidonFrischkorn committed:
Update test-utils.R

- one test was failing because of attributes that were not identical
- i did not know how to address this and thus added a setting to ignore the attributes

This is a test you @venpopov wrote, so maybe you can check this.

GidonFrischkorn committed:
Correct typo in `create_initfun.bmmodel`

GidonFrischkorn committed:
Add minimal tests & minor fix

GidonFrischkorn committed:
Update helpers-inits.R

GidonFrischkorn committed:
Update helpers-model.R

GidonFrischkorn committed:
update `create_initfun` to work with arrays

GidonFrischkorn committed:
Merge branch '286-helpers-extract-stancode' into 285-create_initfun

GidonFrischkorn committed:
Update helpers-inits.R

- resolve dimensions mismatch for `create_initfun` with backend `rstan`

GidonFrischkorn commented:
I think https://github.com/venpopov/bmm/pull/290/commits/4dd618cc8cc238625081d58178fe0398e395c61d has resolved the issue with using the `create_initfun` method with `rstan`. It seems that sometimes a single value is interpreted as a real, but the way some `sd` parameters are declared by `brms` they are set up as vectors.

I found a solution that suggested wrapping the single value in an array explicitly declaring its `dim`.

For now that seems to have done the trick.

GidonFrischkorn committed:
Update helpers-inits.R

- also add the array conversion to vectors of b-parameters with length 1

GidonFrischkorn commented:
Short update:

The `create_initfun` method works well with all three evidence accumulation models (`ddm`, `ezdm`, and `cswald`) that are currently under development (#280, #277, #281 ). In addition, I have also further tested that initial values are now working also with `rstan` as a backend.

I have still not come up with a good idea for tests that does not require to actually fit models. So, if you have an idea I would be happy about your input.

GidonFrischkorn committed:
Merge branch 'develop' into 285-create_initfun

venpopov committed:
Merge branch 'develop' into 285-create_initfun

GidonFrischkorn committed:
Hand NULL in `link_transform` and additional tests

venpopov commented on R/helpers-inits.R:
```diff
@@ -0,0 +1,207 @@
+#' @export
+create_initfun.bmmodel <- function(model, data, formula) {
+  # skip to default method if model object contains no init_ranges
+  # somehow this does not seem to work... only if I do this at the bottom
```
This is because NextMethod is a regular function - whatever is returned by the nextmethod will be returned here. You do not assign the result to anything and the code continues to execute the remaining lines. It's like calling:

```R
if (is.null(model$init_ranges))  {
   create_initfun.sdm(...)
}
# ... rest of the code
```
When NextMethod is at the end, it returns because R implicitly returns the last evaluated argument
So if you want to skip everything in this function, you can explicitly make an early return:

```R
if (is.null(model$init_ranges)) {
    return(NextMethod("create_initfun"))
  }
```

venpopov replied:
which in this case I even simplified to directly return 1 instead of calling a method to return 1. Let me know if there was a different reason to call NextMethod?

Currently it doesn't need to be a method because it only depends on whether the model has init_ranges specified, but not what the model is. If you had in mind that different models might have different create_initfun.* methods, then there will need to be a different logic to route it

venpopov committed:
refactor create_inits function

venpopov commented:
I did some major refactoring to simplify the multiple nested conditionals and eliminate redunancies. All the tests pass, but please check it against your in progress models, in case something I changed was not caught by the tests. Also, see comment above about the method routing logic

venpopov committed:
remove incorrect statement about return value

venpopov commented on R/helpers-inits.R:
```diff
@@ -0,0 +1,156 @@
+      } else {
+        stopif(is.null(nbins) || nbins < 1, "Provide a positive 'nbins' for numeric_strategy='bins'.")
+        x2 <- if (na.rm) x[!is.na(x)] else x
+        if (!length(x2)) {
```
I already simplified the function but is this remaining complexity necessary? You only use the count_term_levels function once with the default. Are there any cases when continuous predictors having bins or multiple terms, unless they are already coded as factors?

GidonFrischkorn replied:
I have to check this more closely to remind myself why I implemented it this way. Would you prefer resolving this before accepting the pull request? Then I will try to have a look at it tomorrow.

GidonFrischkorn commented:
> I did some major refactoring to simplify the multiple nested conditionals and eliminate redunancies. All the tests pass, but please check it against your in progress models, in case something I changed was not caught by the tests.

This looks a lot simpler and a quick test with the current `cswald` implementation did run. The checks with correlations between parameters also ran fine. So from my perspective this should be good for a merge to `develop`.

Have you considered which types of additional tests we could implement to ensure proper functionality?

GidonFrischkorn committed:
Simplify count_term_levels function by removing unused complexity

Remove unused numeric_strategy parameter options ('unique', 'bins'), nbins parameter,
and na.rm parameter from count_term_levels(). The function is only called once with
default parameters, so the additional complexity was unnecessary.

Now the function only handles the actual use case: counting factor levels and treating
continuous predictors as single coefficients. Changed return value from list to scalar
for simplicity.

Addresses PR review comment by @venpopov in #290.

GidonFrischkorn committed:
Fix empty test in test-helpers-init.R

The 'initfun output matches standata dimensions' test was consistently skipped
as 'empty test' because it used an intercept-only model that had no b_
parameters, causing the test expectations inside the conditional loop to never
execute.

Changes:
- Modified test to use a model with predictors (intercept + condition factor)
  to ensure b_ parameters exist
- Fixed dimension checking logic to use Kc_ (centered predictors) for models
  with intercepts, falling back to K_ for models without intercepts
- Added explicit expectation that b_ parameters exist before the loop
- Added informative messages to expectations for better debugging

The test now properly validates that initial values have correct dimensions
matching the Stan model specification.

GidonFrischkorn committed:
Add test for no-intercept model dimension validation

Added a new test to validate that initfun output dimensions match standata
for models with suppressed intercepts (0 + predictor formula).

The test ensures:
- b_ parameters exist for the no-intercept model
- No Intercept_kappa parameter is created (since intercept is suppressed)
- Dimensions match using K_ (total predictors) rather than Kc_ (centered)
  for the suppressed intercept case

This complements the existing test that validates dimension matching for
models with intercepts.

GidonFrischkorn commented:
I have simplified the helper function and agree that typically numerical predictors should always be handled as factors. I have also reworked one of the tests that was skipped before because it was not properly specified.

venpopov committed:
add fallback to unique values for count terms

venpopov reviewed (approved):
Looks good now, thanks. In the future we have see how if at all this needs to be adapted for different models - see my comment above. Currently there is no need for it to be an S3 method, but it's fine to keep it that way
